HTML <script> type=importmap prep (#22300)
mdn/content: files/en-us/web/javascript/guide/modules/index.md

import { name, draw, reportArea, reportPerimeter } from './modules/square.js';
```

You use the {{JSxRef("Statements/import", "import")}} statement, followed by a comma-separated list of the features you want to import wrapped in curly braces, followed by the keyword `from`, followed by the _module specifier_.

The _module specifier_ provides a string that the JavaScript environment can resolve to a path to the module file.

> On a browser this could be a path relative to the site root, which for our `basic-modules` example would be `/js-examples/module-examples/basic-modules`.
> However here we are instead using the dot (`.`) syntax to mean "the current location", followed by the path beyond that to the file we are trying to find. This is much better than writing out the entire relative path each time, as it is shorter, and it makes the URL portable — the example will still work if you move it to a different location in the site hierarchy.

Reviewer: ```suggestion
In a browser, this could be a path relative to the site root, which for our `basic-modules` example would be `/js-examples/module-examples/basic-modules`.
However, here we are instead using the dot (`.`) syntax to mean "the current location", followed by the relative path to the file we are trying to find. This is much better than writing out the entire absolute path each time, as relative paths are shorter and make the URL portable — the example will still work if you move it to a different location in the site hierarchy.
```

You can see such lines in action in [`main.js`](https://github.com/mdn/js-examples/blob/master/module-examples/basic-modules/main.js).

> > **Note:** In some module systems, you can omit the file extension and the leading `/`, `./`, or `../` (e.g. `'modules/square'`). This doesn't work in the browser environment, as that can lead to multiple network roundtrips.

Reviewer: Still mention it, and say you need an import map to make it work? This is referring to node modules which is frankly quite different from "just omitting the leading `./`".

Author: Done!

> **Note:** Although imported features are available in the file, they are read only views of the feature that was exported. You cannot change the variable that was imported, but you can still modify properties similar to `const`. Additionally, these features are imported as live bindings, meaning that they can change in value even if you cannot modify the binding unlike `const`.

> ### Importing bare names

Author: So this is update to modules. This is IMO the main use case of import maps, so here we have the same example again. I think it is worth having this here - it has proper caveats that this requires import maps.

> **Note:** Although imported features are available in the file, they are read only views of the feature that was exported. You cannot change the variable that was imported, but you can still modify properties similar to `const`. Additionally, these features are imported as live bindings, meaning that they can change in value even if you cannot modify the binding unlike `const`.

> ## Importing modules using import maps

Reviewer: General comment: this looks a bit duplicated from the import map doc. I know you've commented about that somewhere already, so if you think it's fine, I'm also good.

Author: I'm OK with it at this point. Partially because I expect you'll be pulling all this apart again at some point soon and you might take a different view.

Above we saw how a browser can import a module using a module specifier that is either an absolute URL, or a relative URL that is resolved using the base URL of the document:

```js
import { name as squareName, draw } from "./shapes/square.js";
import { name as circleName } from "https://example.com/shapes/circle.js";
```

[Import maps](/en-US/docs/Web/HTML/Element/script/type/importmap) allow developers to specify (almost) any text they want in the module specifier; the map provides a corresponding value that will replace the text when the address of the module is resolved.

For example, the import map below defines keys `module/square` and `circle` that have values corresponding to the module specifiers used above.

```html
<script type="importmap">
  {
    "imports": {
      "module/square": "./shapes/square.js",
      "circle": "https://example.com/shapes/circle.js"
    }
  }
</script>
```

> With this import map a script can now use `module/square` and `circle` as module specifiers:

Reviewer: ```suggestion
With this import map, a script can now use `module/square` and `circle` as module specifiers:
```

```js
import { name as squareName, draw } from "module/square";
import { name as circleName } from "circle";
```

> The browser matches the module specifier in the import with a module specifier map key in the import map, in this case `circle` and `module/square`, and replaces matched keys with their associated values.

Reviewer: ```suggestion
The browser matches the module specifier in the import with a module specifier map key in the import map — in this case, `circle` and `module/square` — and replaces matched keys with their associated values.
```

Note that mapped relative address values are first resolved to absolute URL addresses using the [base URL](/en-US/docs/Web/HTML/Element/base) of the document containing the import map.

> Import maps allow modules to be imported using bare module names, similar to how they are imported with Node.js, and can also simulate importing modules from packages, and using modules without file extensions.

Reviewer: ```suggestion
Import maps allow simulating importing modules from packages as if in Node.js by using bare module names, and also allow using modules without file extensions.
```

Author: Not quite - I had three cases  I wanted to cover - bare modules, packages of modules where the module has a file extensions, packages of modules without file extensions. The suggestion doesn't quite work. Let me try make it simpler.

```suggestion
Import maps allow modules to be imported using bare module names (as in Node.js), and can also simulate importing modules from packages, both with and without file extensions.
```

They also allow particular versions of a library to be imported, based on the path of the script that is importing the module.

More generally, import maps let developers write more ergonomic import code, and make it easier to manage the different versions and dependencies of modules used by a site.
This can reduce the effort required to use the same JavaScript libraries in both browser and server.

> ### Importing modules as bare names

Reviewer: I suggest removing this section, because I believe the text in the intro paragraph about using bare module names is self-explanatory already.

Author: I want to keep this for linking and highlighting this, the major use case. I'll make the introduction more interesting so it doesn't just cover this use case.

Author: Yes. I don't think we particularly needed to draw attention to that detail. But to keep you happy I have rewored :-)

In some JavaScript environments you can use bare names for the module specifier.
This works because the environment can resolve module names to a standard location in the file system.
For example, in Node.js you might use the following syntax to import the "square" module.

```js

> import { name, draw, reportArea, reportPerimeter } from 'square';

Reviewer: ```suggestion
import { name, draw, reportArea, reportPerimeter } from "square";
```

```

A bare name will throw a `TypeError` on a browser that does not [support import maps](/en-US/docs/Web/HTML/Element/script/type/importmap#browser_compatibility), because the browser has no way to resolve the module name to a URL.

Using an import map we can provide the information needed by a browser to resolve the URL.
Below you can see a map that defines a `square` module specifier key, which in this case maps to a relative address value.

```html
<script type="importmap">
  {
    "imports": {
      "square": "./shapes/square.js"
    }
  }
</script>
```

With this map we can now use a bare name when we import the module:

```js
import { name as squareName, draw } from "square";
```

### Remapping module paths

Module specifier map entries, where both the specifier key and its associated value have a trailing forward slash (`/`), can be used as a path-prefix.
This allows the remapping of a whole set of import URLs from one location to another.

> It can also be used to emulate working with "packages and modules", such as you might see in the Node ecosystem.

Reviewer: ```suggestion
It can also be used to emulate working with "packages and modules", such as what you might see in the Node ecosystem.
```

Author: This isn't better.

> **Note:** The trailing `/` indicates that the module specifier key can be substituted as _part_ of a module specifier.

> > If this is not present the browser will only match (and substitute) the whole module specifier key.

Reviewer: ```suggestion
> If this is not present, the browser will only match (and substitute) the whole module specifier key.
```

    "lodash/": "/node_modules/lodash-es/"
  }
}
```

With this mapping you can import both the whole "package", using the bare name, and modules within it (using the path mapping):

```js
import _ from "lodash";
import fp from "lodash/fp.js";
```

It is possible to import `fp` above without the `.js` file extension, but you would need to create a bare module specifier key for that file, such as `lodash/fp`, rather than using the path.
This may be reasonable for just one module, but scales poorly if you wish to import many modules.

#### General URL remapping

A module specifier key doesn't have to be path — it can also be an absolute URL (or a URL-like relative path like `./`, `../`, `/`).
This may be useful if you want to remap a module that has absolute paths to a resource with your own local resources.

```json
{
  "imports": {
    "https://www.unpkg.com/moment/": "/node_modules/moment/"
  }
}
```

### Scoped modules for version management

Ecosystems like Node use package managers such as npm to manage modules and their dependencies.
The package manager ensures that each module is separated from other modules and their dependencies.

> As a result, while a complex application might include the same module multiple times, and in several different versions, users do not need to think about this complexity.

Reviewer: ```suggestion
As a result, while a complex application might include the same module multiple times with several different versions in different parts of the module graph, users do not need to think about this complexity.
```

> > **Note:** You can do the same kind of thing using relative paths, but that forces a particular structure on your project, and means that you can't use bare module names, and so on.

Reviewer: ```suggestion
> **Note:** You can also achieve version management using relative paths, but this is subpar because, among other things, this forces a particular structure on your project, and prevents you from using bare module names.
```

Import maps similarly allow you to have multiple versions of dependencies in your application and refer to them using the same module specifier.

> The way this works is that the `scopes` key allows you to provide module specifier maps that will be used depending on the path of the script that is performing the import.

Reviewer: ```suggestion
You implement this with the `scopes` key, which allows you to provide module specifier maps that will be used depending on the path of the module performing the import.
```

The example below demonstrates this.

```json
{
  "imports": {
    "coolmodule": "/node_modules/coolmodule/index.js"
  },
  "scopes": {
    "/node_modules/dependency/": {
      "coolmodule": "/node_modules/some/other/location/coolmodule/index.js"
    }
  }
}
```

With this mapping, if a script with an URL that contains `/node_modules/dependency/` imports `coolmodule`, the version in `/node_modules/some/other/location/coolmodule/index.js` will be used.

> If `coolmodule` is imported from a script with a non-matching scope path, then the "module specifier map" in `imports` will be used instead, mapping to the version in `/node_modules/coolmodule/index.js"`. The map in `imports` would also be used as a fallback if there was no matching specifier in the scoped map.

Reviewer: ```suggestion
The map in `imports` is used as a fallback if there is no matching scope in the scoped map, or the matching scopes don't contain a matching specifier. For example, if `coolmodule` is imported from a script with a non-matching scope path, then the module specifier map in `imports` will be used instead, mapping to the version in `/node_modules/coolmodule/index.js`.
```

Note that the path used to select a scope does not affect how the address is resolved.
The value in the mapped path does not have to have to match the scopes path, and relative paths are still resolved to the base URL of the script that contains the import map.

Just as for module specifier maps, you can have many scope keys, and these may contain overlapping paths.

> If multiple scopes match the referrer URL, then the most specific scope path is checked first (the longest scope key) for a matching specifier.

Reviewer: > Just as for module specifier maps

But the order-by-specificity feature of module specifier maps is not mentioned on this page, while I think it should, because it may not be immediately apparent. In Node, entries in specifier maps are tried by source order, which breaks from the JSON spec.

The browsers will fall back to the next most specific matching scoped path if there is no matching specifier, and so on.
If there is no matching specifier in any of the matching scopes, the browser checks for a match in the module specifier map in the `imports` key.

### Improve caching by mapping away hashed filenames

Script files used by websites often have hashed filenames to simplify caching.
The downside of this approach is that if a module changes, any modules that import it using its hashed filename will also need to be updated/regenerated.
This potentially results in a cascade of updates, which is wasteful of network resources.

Import maps provide a convenient solution to this problem.
Rather than depending on specific hashed filenames, applications and scripts instead depend on an un-hashed version of the module name (address).
An import map like the one below then provides a mapping to the actual script file.

```json
{
  "imports": {
    "main_script": "/node/srcs/application-fg7744e1b.js",
    "dependency_script": "/node/srcs/dependency-3qn7e4b1q.js"
  }
}
```

> If `dependency_script` changes then we have to update the import map to reflect the changed name of the module.
> We don't have to update `main_script` that depends on it, because the `main_script` import code does not change.

Reviewer: ```suggestion
If `dependency_script` changes, then its hash contained in the file name changes as well. In this case, we only need to update the import map to reflect the changed name of the module.
We don't have to update the code of any JavaScript module that depends on it, because the specifier in the import statement does not change.
```

(That's what I read from this)

Author: Yes, much better.

### Dynamic import

Import maps also work with [dynamic imports](#dynamic_module_loading) (see section below).

> This has the same restrictions as for static imports: the import must happen before any other module importing has started, and the module specifier must be resolvable to an absolute URL address for a module.

Reviewer: > the import must happen before any other module importing has started

I don't understand this constraint, and I think that defeats a main purpose of dynamic imports. An example would help.

Also I suggest moving this to the intro paragraph, because this is really orthogonal to all other subsections here, and should be an overarching thing.

Author: >> the import must happen before any other module importing has started

> I don't understand this constraint, and I think that defeats a main purpose of dynamic imports. An example would help.

From [the explainer](https://github.com/WICG/import-maps#import-map-processing)

> Because they affect all imports, any import maps must be present and successfully fetched before any module resolution is done. This means that module graph fetching is blocked on import map fetching.

This is true whether you import a module dynamically or statically, because you don't know what else the module you import is going to do/import.

For similar reasons we only have inline maps definitions - the browser doesn't know what modules it needs until it has parsed the import map. If external maps were allowed then you might fetch the wrong ones unnecessarily while waiting for the map.

Very good point on moving this. I will do this afternoon.

Reviewer: Ah, but the explainer is about dynamically creating/loading an import map, not having an import map and then dynamically importing other modules 😄 That makes more sense. I don't think we are currently discussing the possibility of dynamically-created import maps at all.

Author: Yes. So while we do need to add info up the top about import map being declared before any imports happen, perhaps we don't need this section at all? Just a comment up the top that import maps work with the same for both dynamic and static imports.

Reviewer: It's good if we can mention that import maps should be at the top, yes. And then a comment about dynamic/static imports being the same. That's exactly what I had in mind.

### Feature detection

> You can check support for import maps using the [`HTMLScriptElement.supports()`](http://localhost:5042/en-US/docs/Web/API/HTMLScriptElement/supports) static method (which is itself broadly supported):
> 
> ```js
> if (HTMLScriptElement.supports && HTMLScriptElement.supports("importmap")) {
>   console.log("Browser supports import maps.");
> }
> ```

Reviewer: Similarly, move this up, maybe? (Although less strong opinions on this one, since it's more like an afterthought.) Also:

```suggestion
You can check support for import maps using the [`HTMLScriptElement.supports()`](/en-US/docs/Web/API/HTMLScriptElement/supports) static method (which is itself broadly supported):

```js
if (HTMLScriptElement.supports?.("importmap")) {
  console.log("Browser supports import maps.");
}
```
```

Author: I don't want to move this up because I'm unconvinced of its value. I.e. I think it will be challenging to write code that works sensibly for the case where import maps are supported and are not supported.

In this case the browser will select the most specific (longest) matching module specifier key.

Import maps allow modules to be imported using bare module names (as in Node.js), and can also simulate importing modules from packages, both with and without file extensions.
While not shown above, they also allow particular versions of a library to be imported, based on the path of the script that is importing the module.
Generally they let developers write more ergonomic import code, and make it easier to manage the different versions and dependencies of modules used by a site.
This can reduce the effort required to use the same JavaScript libraries in both browser and server.

The following sections expand on the various features outlined above.

### Feature detection

You can check support for import maps using the [`HTMLScriptElement.supports()`](/en-US/docs/Web/API/HTMLScriptElement/supports) static method (which is itself broadly supported):

```js
if (HTMLScriptElement.supports?.("importmap")) {
  console.log("Browser supports import maps.");
}
```

### Importing modules as bare names

> In some JavaScript environments you can use bare names for the module specifier.
> This works because the environment can resolve module names to a standard location in the file system.
> For example, in Node.js you might use the following syntax to import the "square" module.

Author: ```suggestion
In some JavaScript environments, such as Node.js, you can use bare names for the module specifier.
This works because the environment can resolve module names to a standard location in the file system.
For example, you might use the following syntax to import the "square" module.
```

While not shown above, they also allow particular versions of a library to be imported, based on the path of the script that is importing the module.
Generally they let developers write more ergonomic import code, and make it easier to manage the different versions and dependencies of modules used by a site.
This can reduce the effort required to use the same JavaScript libraries in both browser and server.

The following sections expand on the various features outlined above.

### Feature detection

You can check support for import maps using the [`HTMLScriptElement.supports()`](/en-US/docs/Web/API/HTMLScriptElement/supports) static method (which is itself broadly supported):

```js
if (HTMLScriptElement.supports?.("importmap")) {
  console.log("Browser supports import maps.");
}
```

### Importing modules as bare names

In some JavaScript environments you can use bare names for the module specifier.
This works because the environment can resolve module names to a standard location in the file system.
For example, in Node.js you might use the following syntax to import the "square" module.

```js
import { name, draw, reportArea, reportPerimeter } from "square";
```

> A bare name will throw a `TypeError` on a browser that does not [support import maps](/en-US/docs/Web/HTML/Element/script/type/importmap#browser_compatibility), because the browser has no way to resolve the module name to a URL.

Reviewer: It still throws if the browser supports import maps but there's no matching specifier key. I think the crucial thing is really that you have to make sure it resolves to something.

> A bare name will throw a `TypeError` on a browser that does not [support import maps](/en-US/docs/Web/HTML/Element/script/type/importmap#browser_compatibility), because the browser has no way to resolve the module name to a URL.
> 
> Using an import map we can provide the information needed by a browser to resolve the URL.
> Below you can see a map that defines a `square` module specifier key, which in this case maps to a relative address value.

Author: ```suggestion
To use bare names on a browser you need an import map, which provides the information needed by the browser to resolve module specifiers to URLs (JavaScript will throw a `TypeError` if it attempts to import a module specifier that can't be resolved to a module location).

Below you can see a map that defines a `square` module specifier key, which in this case maps to a relative address value.
```
